### PR TITLE
feat: nix for building Semgrep!

### PR DESCRIPTION
**Pull request by ajbt200128**
TL;DR; Nix is a dependency manager. This PR adds 2 files to support it and makes it so we autobump our semgrep.opam version. This will make it so it takes only one command to have a fully functional correct and near identical development environment across any OS, instead of the 10-20 commands it takes right now. It's totally opt-in and will not impact anyones work flow.

I can separate out the version bump fix into a separate PR if people want.
# What is this PR?
## What is Nix? 
Nix is a package dependency manager for reproducible and correct builds. Nix is structured around the concept that each dependency is a function of its build dependencies. For example a hello world program takes in GCC and make as function arguments, and spits out a hello world binary. What makes this reproducible is that nix treats files as pointers, and produces a closure over all files in a build system. For example if the hello world depends on a C library, then your nix build of hello world won't complete unless you explicitly declare said library in your nix configuration. Since all build tools made with nix are built this way, nix knows exactly what any dependency needs to build, and to run, meaning *all builds are easily reproducible, and are not affected by what you may or may not have installed on your system*.

## Why should I care?
 Our OCaml code is (mostly) correct because that's a focus of the language itself. But it relies on a lot of C code and external dependencies, like `libev`, `libcurl`, `tree-sitter` etc. This means that if someone wants to build and contribute to Semgrep, they must install these dependencies, and hope that their versions is compatible. For example if you install the ocaml packages needed for osemgrep, and then install libev, things won't work, since the lwt package needs libev when its initially installed. If you're on mac, you also have to tell opam where libev is before installing lwt. So our OCaml is correct, but only if you can build it, and only if you build it with the right dependencies.

By using nix, we can declare all these dependencies explicitly, and then anyone across any *nix system can easily build Semgrep with only one command! What's even better, is for regular contributors, these dependencies will auto update and rebuild whenever a new dependency is added. So if someone adds a dependency for lwt, libev will automatically be pulled in with 0 thought from the contributor, and no need to run `make dev-setup` or anything similar. Since all of the dependencies are reproducible, they're almost identical no matter what system you're on, so if someone runs into a bug in developing, you can have more confidence in eliminating a system difference as the source of the bug.

Another bonus is that it's easy to know what version of a dependency you're on, e.g. what version of tree-sitter, or what version of OCaml. You can see how this also would be good for security :).

There's a lot more places this can simplify things for us, like CI and releases. See future work for more details.

Finally, nix configurations are written in a functional programming language. Don't you want to declare dependencies in a functional language? 

## Who uses Nix?
Not convinced? Here are some other notable OCaml projects using nix:

- [dune](https://github.com/ocaml/dune)
- [merlin](https://github.com/ocaml/merlin)
- [ocaml-lsp](https://github.com/ocaml/ocaml-lsp)
- [ocaml-re](https://github.com/ocaml/ocaml-re)

There's also a ton of projects outside the OCaml world using nix too. Nix is battle tested, and has a huge community, with support for almost anything you can think of. There's even someone packaging Semgrep for nix.

## Try it out
If you want to see what it's like to use nix, get a clean Linux or macOS box. Then [install nix](https://github.com/DeterminateSystems/nix-installer) with flake support. Finally run `nix develop`. Now you can run `make core` or any make targets in the CLI. That's one command to always be able to build Semgrep vs the 10-20 needed right now, if you're lucky, and then 2-3 every once in awhile to keep up to date.

If you want to reproduce and help with a bug in someones branch, you can just run `nix develop` and `make test` to instantly have the same exact environment they ran into the bug in. If you're really unsure about if it's a system bug, you can run `nix develop -i` to exclude any non nix dependencies completely. 

If you don't use bash run `nix develop -c $SHELL` to get an environment with your shell.

If you want to just build semgrep and run it you can do `nix run ".?submodules=1#"` for semgrep or `nix run ".?submodules=1#<target>"` where target is semgrep, pysemgrep, osemgrep, or semgrep-core.

## What's the catch?
There's not a real big catch to using nix, except development time to set it up, which this PR just did. A few small catches are that nix can be a little slow on a first build, but no more than say `make dev-setup`, and that it uses a decent amount of storage space, but again, not much more than any other package manager. 

## Contributing/Maintennance
The maintenance for nix is super low. It automatically pulls in any dependency declared in our semgrep.opam file, which then pulls in any non ocaml dependencies those dependencies rely on. If someone adds a new build tool or python package, they will have to add the dependency to `flake.nix`. This means you have to read some comments then add the name of the package to an array somewhere, super easy. If you're interested in doing anything more complex, see the future reading section.

# Future work
## Some issues with our current usage of Dune and Opam
Right now our usage of dune is a little incorrect. In every language semgrep tree-sitter dune file (i.e. bash/tree-sitter/semgrep-bash), we don't pass the standard c/c++ flags to compile the C files. This means that the compiler provided by nix doesn't know if a c++ file is a c++ file, and fails to include the standard c++ library. There's an ongoing PR to fix this logic for nix,  but really we should add ":standard" to the list of c++ flags in these dune files. There's a patch for now in the nix configuration but it's a bit ugly.

Another issue with our opam usage is we don't declare dependencies on libraries provided by semgrep. I.e. we depend on commons, but don't declare that dependency in semgrep.opam. This breaks some autoresolving things in nix, which are patched for now, but would be nice if we fixed our usage of opam.

## CI testing
If people like this tool, at some point it'd be great to add a CI workflow that ensures all PRs work even with nix. But let's see how this goes first.

## Improving our CI
Where nix would smooth things over for us even more is simplifying our CI. Right now we have to guess and hope that our CI has the same dependencies as what's on our system, and it can be a fight to add any new dependencies, like libcurl. Not to mention our workflows differ by system, and it's a pain to make sure they all have the right package names. If we use nix in CI, all we need is to install nix on the target OS/architecture, then run nix build and it can run all build processes, and all e2e tests. Super simple workflow. It can even build the wheels, and static versions of Semgrep we'd want!

# More reading
Interested in reading more about nix? Here are some hand sources
- [What is Nix](https://shopify.engineering/what-is-nix) - short intro to nix by Shopify Engineering
- [Nix Pills](https://nixos.org/guides/nix-pills/) -long form intro to nix
- [Nix Flakes](https://zero-to-nix.com/concepts/flakes) - intro to the format we use to configure nix
- [opam-nix](https://www.tweag.io/blog/2023-02-16-opam-nix/) - how we nixify opam deps
- [Nix PhD Thesis](https://edolstra.github.io/pubs/phd-thesis.pdf) - Nix creator's PhD thesis on Nix. ~275 pages but really approachable

### CODE REVIEW

File: .envrc [hunk 1]
```diff
@@ -0,0 +1 @@
+use flake
```
Comment (aryx) [inline .envrc:4]: need a comment explaining what this file is about, maybe with a link to the reference doc

File: .dockerignore [hunk 2]
```diff
@@ -202,5 +202,9 @@ cli/src/semgrep/bin/*
 # Artifacts from CI
 **/scripts/release/release_body.txt
 
+# nix
+**/result
+# direnv
```
Comment (aryx) [inline .dockerignore:207]: what is direnv? If it's part of nix, then maybe can just have the # nix and remove # direnv.

File: flake.nix [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+        # pysemgrep
+        pysemgrep = with pythonPackages; buildPythonApplication {
+          # thanks to @06kellyjac
+          pname = "semgrep";
+          inherit (osemgrep) version;
+          src = ./cli;
+          # TODO checks
+          doCheck = false;
+
+          propagatedBuildInputs = [
```
Comment (aryx) [inline flake.nix:297]: maybe add some coupling: comment in this file and the related pypi file so if
one adds a python dependency, we know we also need to update this file.

**Review by aryx — APPROVED** [review]
LGTM.
need to fix the CI failures though.

File: flake.nix [hunk 4]
```diff
@@ -0,0 +1,210 @@
+# you found our nix flake!
```
Comment (aryx) [inline flake.nix:1]: Maybe put most of the PR description in this file as a big comment.
Also why it's called flake.nix instead of semgrep.nix?

**Review by aryx — CHANGES_REQUESTED** [review]
more comments in the files instead of the PR description.
otherwise LGTM.

File: .dockerignore (hunk 2 above)
Comment (ajbt200128) [reply .dockerignore:207]: direnv activates arbitrary hooks when entering a directory. Added it so we can start a nix shell when entering semgrep, but can be used for other things, so will keep it separate

File: .envrc (hunk 1 above)
Comment (ajbt200128) [reply .envrc:4]: will add

File: flake.nix (hunk 3 above)
Comment (ajbt200128) [reply flake.nix:297]: added

File: flake.nix (hunk 4 above)
Comment (ajbt200128) [reply flake.nix:1]: added it to the file. flake.nix is the name that most nix related things expect. Since a flake can export multiple programs and libraries (i.e. semgrep, osemgrep etc.) it's also better to call it one thing

File: .envrc (hunk 1 above)
Comment (aryx) [reply .envrc:4]: the comment is talking about direnv, but the code is talking about a flake ...

File: flake.nix [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+        # pysemgrep
+        pysemgrep = with pythonPackages; buildPythonApplication {
+          # thanks to @06kellyjac
+          pname = "semgrep";
+          inherit (osemgrep) version;
+          src = ./cli;
+          # TODO checks
+          doCheck = false;
+
+          # coupling: anything added to the pysemgrep setup.py should be added here
```
Comment (aryx) [inline flake.nix:296]: the comment should really be in setup.py instead since that's most likely what people will modify first.

**Review by aryx — APPROVED** [review]
Using 80columns would be even better :)

File: flake.nix (hunk 5 above)
Comment (ajbt200128) [reply flake.nix:296]: will add in both places
